File: mini-agent-harness/typescript/agent/releaseControlPlane.ts
export type ProtocolRange = Readonly<{ min: number; max: number }>

export type ReleaseManifest = Readonly<{
  releaseId: string
  binaryVersion: string
  protocol: ProtocolRange
  readableSchemaVersions: readonly number[]
  writeSchemaVersion: number
  policyRevision: number
  featureRevision: number
}>

export type RolloutStage = Readonly<{
  name: string
  trafficPercent: number
  minSamples: number
}>

export type SloPolicy = Readonly<{
  minSuccessRate: number
  maxP95LatencyMs: number
  maxObserverDropRate: number
  maxUnknownCostRate: number
}>

export type SliWindow = Readonly<{
  sampleCount: number
  successRate: number
  p95LatencyMs: number
  observerDropRate: number
  unknownCostRate: number
}>

export type WorkerDependencies = Readonly<{
  transcriptStore: boolean
  workQueue: boolean
  policySource: boolean
  quotaStore: boolean
}>

export type WorkerRegistration = Readonly<{
  workerId: string
  releaseId: string
  protocolVersion: number
  readableSchemaVersions: readonly number[]
  writeSchemaVersion: number
  policyRevision: number
  featureRevision: number
  dependencies: WorkerDependencies
}>

export type WorkerStatus = 'ready' | 'draining' | 'drained' | 'unhealthy'

export type WorkerReport = Readonly<{
  workerId: string
  releaseId: string
  status: WorkerStatus
  activeWorkCount: number
  protocolVersion: number
  policyRevision: number
  featureRevision: number
}>

export type ReleaseReport = Readonly<{
  revision: number
  activeReleaseId: string
  candidateReleaseId?: string
  previousReleaseId?: string
  rolloutState: 'idle' | 'prepared' | 'canary' | 'promoted' | 'rolled_back'
  stageName?: string
  trafficPercent: number
  readyActiveWorkers: number
  readyCandidateWorkers: number
  drainingWorkers: number
  activeWorkCount: number
}>

export class StaleReleaseRevisionError extends Error {}
export class ReleaseCompatibilityError extends Error {}
export class WorkerCompatibilityError extends Error {}
export class ReleaseStateError extends Error {}
export class SloViolationError extends Error {}
export class WorkerStateError extends Error {}

type WorkerState = {
  readonly registration: WorkerRegistration
  status: WorkerStatus
  activeWorkCount: number
}

type Rollout = {
  readonly stages: readonly RolloutStage[]
  stageIndex: number
  state: 'prepared' | 'canary'
}

/**
 * In-process release state machine. Real routing, durable leases, orchestration
 * and database migration are adapter responsibilities.
 */
export class ReleaseController {
  #revision = 1
  #active: ReleaseManifest
  #candidate: ReleaseManifest | undefined
  #previous: ReleaseManifest | undefined
  #rollout: Rollout | undefined
  #terminalState: 'idle' | 'promoted' | 'rolled_back' = 'idle'
  readonly #workers = new Map<string, WorkerState>()
  readonly #slo: SloPolicy

  constructor(active: ReleaseManifest, slo: SloPolicy) {
    validateManifest(active)
    validateSlo(slo)
    this.#active = freeze(active)
    this.#slo = freeze(slo)
  }

  prepare(
    expectedRevision: number,
    candidate: ReleaseManifest,
    stages: readonly RolloutStage[],
  ): ReleaseReport {
    this.#expectRevision(expectedRevision)
    if (this.#candidate || this.#rollout) throw new ReleaseStateError('a rollout is already active')
    validateManifest(candidate)
    validateStages(stages)
    this.#validateReleaseCompatibility(this.#active, candidate)
    this.#candidate = freeze(candidate)
    this.#rollout = { stages: freeze(stages), stageIndex: -1, state: 'prepared' }
    this.#terminalState = 'idle'
    this.#revision++
    return this.report()
  }

  registerWorker(registration: WorkerRegistration): WorkerReport {
    if (this.#workers.has(registration.workerId)) {
      throw new WorkerStateError(`duplicate worker id: ${registration.workerId}`)
    }
    const manifest = this.#manifestFor(registration.releaseId)
    validateWorker(registration, manifest)
    const healthy = Object.values(registration.dependencies).every(Boolean)
    const state: WorkerState = {
      registration: freeze(registration),
      status: healthy ? 'ready' : 'unhealthy',
      activeWorkCount: 0,
    }
    this.#workers.set(registration.workerId, state)
    return workerReport(state)
  }

  beginCanary(expectedRevision: number): ReleaseReport {
    this.#expectRevision(expectedRevision)
    if (!this.#candidate || !this.#rollout || this.#rollout.state !== 'prepared') {
      throw new ReleaseStateError('no prepared rollout')
    }
    if (this.#readyWorkers(this.#candidate.releaseId) === 0) {
      throw new ReleaseStateError('candidate has no ready worker')
    }
    if (this.#readyWorkers(this.#active.releaseId) === 0) {
      throw new ReleaseStateError('active release has no ready worker')
    }
    this.#rollout.stageIndex = 0
    this.#rollout.state = 'canary'
    this.#revision++
    return this.report()
  }

  selectRelease(stableBucket: number): string {
    if (!Number.isInteger(stableBucket) || stableBucket < 0 || stableBucket > 99) {
      throw new Error('stable bucket must be an integer in [0, 99]')
    }
    if (!this.#candidate || !this.#rollout || this.#rollout.state !== 'canary') {
      return this.#active.releaseId
    }
    const stage = this.#rollout.stages[this.#rollout.stageIndex]!
    return stableBucket < stage.trafficPercent
      ? this.#candidate.releaseId
      : this.#active.releaseId
  }

  advance(expectedRevision: number, window: SliWindow): ReleaseReport {
    this.#expectRevision(expectedRevision)
    if (!this.#candidate || !this.#rollout || this.#rollout.state !== 'canary') {
      throw new ReleaseStateError('no canary rollout')
    }
    const stage = this.#rollout.stages[this.#rollout.stageIndex]!
    assertSlo(window, stage, this.#slo)
    if (this.#rollout.stageIndex < this.#rollout.stages.length - 1) {
      this.#rollout.stageIndex++
    } else {
      this.#previous = this.#active
      this.#active = this.#candidate
      this.#candidate = undefined
      this.#rollout = undefined
      this.#terminalState = 'promoted'
    }
    this.#revision++
    return this.report()
  }

  rollback(expectedRevision: number): ReleaseReport {
    this.#expectRevision(expectedRevision)
    if (this.#candidate) {
      this.#markReleaseDraining(this.#candidate.releaseId)
      this.#candidate = undefined
      this.#rollout = undefined
      this.#terminalState = 'rolled_back'
    } else if (this.#previous) {
      if (!this.#previous.readableSchemaVersions.includes(this.#active.writeSchemaVersion)) {
        throw new ReleaseCompatibilityError('previous release cannot read active write schema')
      }
      const rolledRelease = this.#active
      this.#active = this.#previous
      this.#previous = rolledRelease
      this.#markReleaseDraining(rolledRelease.releaseId)
      this.#terminalState = 'rolled_back'
    } else {
      throw new ReleaseStateError('no candidate or previous release to roll back')
    }
    this.#revision++
    return this.report()
  }

  acquireWork(workerId: string, selectedReleaseId: string): WorkerReport {
    const state = this.#worker(workerId)
    if (state.status !== 'ready') throw new WorkerStateError(`worker is not ready: ${workerId}`)
    if (state.registration.releaseId !== selectedReleaseId) {
      throw new WorkerStateError(`worker release does not match selected release: ${workerId}`)
    }
    state.activeWorkCount++
    return workerReport(state)
  }

  completeWork(workerId: string): WorkerReport {
    const state = this.#worker(workerId)
    if (state.activeWorkCount < 1) throw new WorkerStateError(`worker has no active work: ${workerId}`)
    state.activeWorkCount--
    if (state.status === 'draining' && state.activeWorkCount === 0) state.status = 'drained'
    return workerReport(state)
  }

  drainWorker(workerId: string): WorkerReport {
    const state = this.#worker(workerId)
    if (state.status === 'drained') return workerReport(state)
    state.status = state.activeWorkCount === 0 ? 'drained' : 'draining'
    return workerReport(state)
  }

  worker(workerId: string): WorkerReport {
    return workerReport(this.#worker(workerId))
  }

  report(): ReleaseReport {
    const stage = this.#rollout && this.#rollout.stageIndex >= 0
      ? this.#rollout.stages[this.#rollout.stageIndex]
      : undefined
    const rolloutState = this.#rollout?.state ?? this.#terminalState
    return Object.freeze({
      revision: this.#revision,
      activeReleaseId: this.#active.releaseId,
      ...(this.#candidate ? { candidateReleaseId: this.#candidate.releaseId } : {}),
      ...(this.#previous ? { previousReleaseId: this.#previous.releaseId } : {}),
      rolloutState,
      ...(stage ? { stageName: stage.name } : {}),
      trafficPercent: stage?.trafficPercent ?? 0,
      readyActiveWorkers: this.#readyWorkers(this.#active.releaseId),
      readyCandidateWorkers: this.#candidate ? this.#readyWorkers(this.#candidate.releaseId) : 0,
      drainingWorkers: [...this.#workers.values()].filter(worker => worker.status === 'draining').length,
      activeWorkCount: [...this.#workers.values()].reduce((sum, worker) => sum + worker.activeWorkCount, 0),
    })
  }

  #validateReleaseCompatibility(active: ReleaseManifest, candidate: ReleaseManifest): void {
    if (active.releaseId === candidate.releaseId) throw new ReleaseCompatibilityError('release id must change')
    if (!rangesOverlap(active.protocol, candidate.protocol)) {
      throw new ReleaseCompatibilityError('protocol ranges do not overlap')
    }
    if (!candidate.readableSchemaVersions.includes(active.writeSchemaVersion)) {
      throw new ReleaseCompatibilityError('candidate cannot read active write schema')
    }
    if (!active.readableSchemaVersions.includes(candidate.writeSchemaVersion)) {
      throw new ReleaseCompatibilityError('rollback release cannot read candidate write schema')
    }
    if (active.policyRevision !== candidate.policyRevision) {
      throw new ReleaseCompatibilityError('binary canary cannot split policy revision')
    }
  }

  #manifestFor(releaseId: string): ReleaseManifest {
    if (this.#active.releaseId === releaseId) return this.#active
    if (this.#candidate?.releaseId === releaseId) return this.#candidate
    if (this.#previous?.releaseId === releaseId) return this.#previous
    throw new WorkerCompatibilityError(`unknown release: ${releaseId}`)
  }

  #readyWorkers(releaseId: string): number {
    return [...this.#workers.values()].filter(
      worker => worker.registration.releaseId === releaseId && worker.status === 'ready',
    ).length
  }

  #markReleaseDraining(releaseId: string): void {
    for (const worker of this.#workers.values()) {
      if (worker.registration.releaseId === releaseId && worker.status === 'ready') {
        worker.status = worker.activeWorkCount === 0 ? 'drained' : 'draining'
      }
    }
  }

  #worker(workerId: string): WorkerState {
    const worker = this.#workers.get(workerId)
    if (!worker) throw new WorkerStateError(`unknown worker: ${workerId}`)
    return worker
  }

  #expectRevision(expected: number): void {
    if (expected !== this.#revision) {
      throw new StaleReleaseRevisionError(`stale release revision ${expected}; current=${this.#revision}`)
    }
  }
}

function validateManifest(manifest: ReleaseManifest): void {
  requireText(manifest.releaseId, 'releaseId')
  requireText(manifest.binaryVersion, 'binaryVersion')
  validatePositiveInteger(manifest.protocol.min, 'protocol.min')
  validatePositiveInteger(manifest.protocol.max, 'protocol.max')
  if (manifest.protocol.min > manifest.protocol.max) throw new Error('protocol range is inverted')
  if (manifest.readableSchemaVersions.length === 0) throw new Error('at least one readable schema is required')
  for (const version of manifest.readableSchemaVersions) validatePositiveInteger(version, 'readable schema')
  validatePositiveInteger(manifest.writeSchemaVersion, 'writeSchemaVersion')
  validatePositiveInteger(manifest.policyRevision, 'policyRevision')
  validatePositiveInteger(manifest.featureRevision, 'featureRevision')
  if (!manifest.readableSchemaVersions.includes(manifest.writeSchemaVersion)) {
    throw new Error('release must read its own write schema')
  }
}

function validateWorker(worker: WorkerRegistration, manifest: ReleaseManifest): void {
  requireText(worker.workerId, 'workerId')
  if (worker.protocolVersion < manifest.protocol.min || worker.protocolVersion > manifest.protocol.max) {
    throw new WorkerCompatibilityError('worker protocol is outside release range')
  }
  if (worker.writeSchemaVersion !== manifest.writeSchemaVersion) {
    throw new WorkerCompatibilityError('worker write schema does not match manifest')
  }
  if (!manifest.readableSchemaVersions.every(version => worker.readableSchemaVersions.includes(version))) {
    throw new WorkerCompatibilityError('worker cannot read every manifest schema')
  }
  if (worker.policyRevision !== manifest.policyRevision) {
    throw new WorkerCompatibilityError('worker policy revision does not match manifest')
  }
  if (worker.featureRevision !== manifest.featureRevision) {
    throw new WorkerCompatibilityError('worker feature revision does not match manifest')
  }
}

function validateStages(stages: readonly RolloutStage[]): void {
  if (stages.length === 0) throw new Error('rollout needs at least one stage')
  let previous = 0
  for (const stage of stages) {
    requireText(stage.name, 'stage.name')
    if (!Number.isInteger(stage.trafficPercent) || stage.trafficPercent <= previous || stage.trafficPercent > 100) {
      throw new Error('traffic stages must increase to at most 100')
    }
    validatePositiveInteger(stage.minSamples, 'stage.minSamples')
    previous = stage.trafficPercent
  }
  if (previous !== 100) throw new Error('last rollout stage must be 100 percent')
}

function validateSlo(slo: SloPolicy): void {
  validateRatio(slo.minSuccessRate, 'minSuccessRate')
  validateRatio(slo.maxObserverDropRate, 'maxObserverDropRate')
  validateRatio(slo.maxUnknownCostRate, 'maxUnknownCostRate')
  if (!Number.isFinite(slo.maxP95LatencyMs) || slo.maxP95LatencyMs <= 0) {
    throw new Error('maxP95LatencyMs must be positive')
  }
}

function assertSlo(window: SliWindow, stage: RolloutStage, slo: SloPolicy): void {
  if (
    window.sampleCount < stage.minSamples
    || window.successRate < slo.minSuccessRate
    || window.p95LatencyMs > slo.maxP95LatencyMs
    || window.observerDropRate > slo.maxObserverDropRate
    || window.unknownCostRate > slo.maxUnknownCostRate
  ) {
    throw new SloViolationError(`SLO rejected rollout stage: ${stage.name}`)
  }
}

function rangesOverlap(left: ProtocolRange, right: ProtocolRange): boolean {
  return left.min <= right.max && right.min <= left.max
}

function workerReport(state: WorkerState): WorkerReport {
  return Object.freeze({
    workerId: state.registration.workerId,
    releaseId: state.registration.releaseId,
    status: state.status,
    activeWorkCount: state.activeWorkCount,
    protocolVersion: state.registration.protocolVersion,
    policyRevision: state.registration.policyRevision,
    featureRevision: state.registration.featureRevision,
  })
}

function validateRatio(value: number, name: string): void {
  if (!Number.isFinite(value) || value < 0 || value > 1) throw new Error(`${name} must be in [0, 1]`)
}

function validatePositiveInteger(value: number, name: string): void {
  if (!Number.isInteger(value) || value < 1) throw new Error(`${name} must be a positive integer`)
}

function requireText(value: string, name: string): void {
  if (value.trim() === '') throw new Error(`${name} must not be empty`)
}

function freeze<T>(value: T): T {
  const clone = structuredClone(value)
  return deepFreeze(clone)
}

function deepFreeze<T>(value: T): T {
  if (value && typeof value === 'object' && !Object.isFrozen(value)) {
    Object.freeze(value)
    for (const child of Object.values(value)) deepFreeze(child)
  }
  return value
}

File: mini-agent-harness/python/release_control_plane.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Literal, Mapping


@dataclass(frozen=True)
class ProtocolRange:
    minimum: int
    maximum: int


@dataclass(frozen=True)
class ReleaseManifest:
    release_id: str
    binary_version: str
    protocol: ProtocolRange
    readable_schema_versions: tuple[int, ...]
    write_schema_version: int
    policy_revision: int
    feature_revision: int


@dataclass(frozen=True)
class RolloutStage:
    name: str
    traffic_percent: int
    min_samples: int


@dataclass(frozen=True)
class SloPolicy:
    min_success_rate: float
    max_p95_latency_ms: float
    max_observer_drop_rate: float
    max_unknown_cost_rate: float


@dataclass(frozen=True)
class SliWindow:
    sample_count: int
    success_rate: float
    p95_latency_ms: float
    observer_drop_rate: float
    unknown_cost_rate: float


@dataclass(frozen=True)
class WorkerDependencies:
    transcript_store: bool
    work_queue: bool
    policy_source: bool
    quota_store: bool

    def ready(self) -> bool:
        return all(
            (self.transcript_store, self.work_queue, self.policy_source, self.quota_store)
        )


@dataclass(frozen=True)
class WorkerRegistration:
    worker_id: str
    release_id: str
    protocol_version: int
    readable_schema_versions: tuple[int, ...]
    write_schema_version: int
    policy_revision: int
    feature_revision: int
    dependencies: WorkerDependencies


WorkerStatus = Literal["ready", "draining", "drained", "unhealthy"]


@dataclass(frozen=True)
class WorkerReport:
    worker_id: str
    release_id: str
    status: WorkerStatus
    active_work_count: int
    protocol_version: int
    policy_revision: int
    feature_revision: int


@dataclass(frozen=True)
class ReleaseReport:
    revision: int
    active_release_id: str
    candidate_release_id: str | None
    previous_release_id: str | None
    rollout_state: Literal["idle", "prepared", "canary", "promoted", "rolled_back"]
    stage_name: str | None
    traffic_percent: int
    ready_active_workers: int
    ready_candidate_workers: int
    draining_workers: int
    active_work_count: int


class StaleReleaseRevisionError(RuntimeError):
    pass


class ReleaseCompatibilityError(ValueError):
    pass


class WorkerCompatibilityError(ValueError):
    pass


class ReleaseStateError(RuntimeError):
    pass


class SloViolationError(RuntimeError):
    pass


class WorkerStateError(RuntimeError):
    pass


@dataclass
class _WorkerState:
    registration: WorkerRegistration
    status: WorkerStatus
    active_work_count: int = 0


@dataclass
class _Rollout:
    stages: tuple[RolloutStage, ...]
    stage_index: int = -1
    state: Literal["prepared", "canary"] = "prepared"


class ReleaseController:
    """In-process reference state machine, not a deployment orchestrator."""

    def __init__(self, active: ReleaseManifest, slo: SloPolicy) -> None:
        _validate_manifest(active)
        _validate_slo(slo)
        self._revision = 1
        self._active = active
        self._candidate: ReleaseManifest | None = None
        self._previous: ReleaseManifest | None = None
        self._rollout: _Rollout | None = None
        self._terminal_state: Literal["idle", "promoted", "rolled_back"] = "idle"
        self._workers: dict[str, _WorkerState] = {}
        self._slo = slo

    def prepare(
        self,
        expected_revision: int,
        candidate: ReleaseManifest,
        stages: tuple[RolloutStage, ...],
    ) -> ReleaseReport:
        self._expect_revision(expected_revision)
        if self._candidate or self._rollout:
            raise ReleaseStateError("a rollout is already active")
        _validate_manifest(candidate)
        _validate_stages(stages)
        self._validate_release_compatibility(self._active, candidate)
        self._candidate = candidate
        self._rollout = _Rollout(stages)
        self._terminal_state = "idle"
        self._revision += 1
        return self.report()

    def register_worker(self, registration: WorkerRegistration) -> WorkerReport:
        if registration.worker_id in self._workers:
            raise WorkerStateError(f"duplicate worker id: {registration.worker_id}")
        manifest = self._manifest_for(registration.release_id)
        _validate_worker(registration, manifest)
        state = _WorkerState(
            registration,
            "ready" if registration.dependencies.ready() else "unhealthy",
        )
        self._workers[registration.worker_id] = state
        return _worker_report(state)

    def begin_canary(self, expected_revision: int) -> ReleaseReport:
        self._expect_revision(expected_revision)
        if not self._candidate or not self._rollout or self._rollout.state != "prepared":
            raise ReleaseStateError("no prepared rollout")
        if self._ready_workers(self._candidate.release_id) == 0:
            raise ReleaseStateError("candidate has no ready worker")
        if self._ready_workers(self._active.release_id) == 0:
            raise ReleaseStateError("active release has no ready worker")
        self._rollout.stage_index = 0
        self._rollout.state = "canary"
        self._revision += 1
        return self.report()

    def select_release(self, stable_bucket: int) -> str:
        if not isinstance(stable_bucket, int) or not 0 <= stable_bucket <= 99:
            raise ValueError("stable bucket must be an integer in [0, 99]")
        if not self._candidate or not self._rollout or self._rollout.state != "canary":
            return self._active.release_id
        stage = self._rollout.stages[self._rollout.stage_index]
        return (
            self._candidate.release_id
            if stable_bucket < stage.traffic_percent
            else self._active.release_id
        )

    def advance(self, expected_revision: int, window: SliWindow) -> ReleaseReport:
        self._expect_revision(expected_revision)
        if not self._candidate or not self._rollout or self._rollout.state != "canary":
            raise ReleaseStateError("no canary rollout")
        stage = self._rollout.stages[self._rollout.stage_index]
        _assert_slo(window, stage, self._slo)
        if self._rollout.stage_index < len(self._rollout.stages) - 1:
            self._rollout.stage_index += 1
        else:
            self._previous = self._active
            self._active = self._candidate
            self._candidate = None
            self._rollout = None
            self._terminal_state = "promoted"
        self._revision += 1
        return self.report()

    def rollback(self, expected_revision: int) -> ReleaseReport:
        self._expect_revision(expected_revision)
        if self._candidate:
            self._mark_release_draining(self._candidate.release_id)
            self._candidate = None
            self._rollout = None
            self._terminal_state = "rolled_back"
        elif self._previous:
            if self._active.write_schema_version not in self._previous.readable_schema_versions:
                raise ReleaseCompatibilityError(
                    "previous release cannot read active write schema"
                )
            rolled_release = self._active
            self._active = self._previous
            self._previous = rolled_release
            self._mark_release_draining(rolled_release.release_id)
            self._terminal_state = "rolled_back"
        else:
            raise ReleaseStateError("no candidate or previous release to roll back")
        self._revision += 1
        return self.report()

    def acquire_work(self, worker_id: str, selected_release_id: str) -> WorkerReport:
        state = self._worker(worker_id)
        if state.status != "ready":
            raise WorkerStateError(f"worker is not ready: {worker_id}")
        if state.registration.release_id != selected_release_id:
            raise WorkerStateError(
                f"worker release does not match selected release: {worker_id}"
            )
        state.active_work_count += 1
        return _worker_report(state)

    def complete_work(self, worker_id: str) -> WorkerReport:
        state = self._worker(worker_id)
        if state.active_work_count < 1:
            raise WorkerStateError(f"worker has no active work: {worker_id}")
        state.active_work_count -= 1
        if state.status == "draining" and state.active_work_count == 0:
            state.status = "drained"
        return _worker_report(state)

    def drain_worker(self, worker_id: str) -> WorkerReport:
        state = self._worker(worker_id)
        if state.status != "drained":
            state.status = "drained" if state.active_work_count == 0 else "draining"
        return _worker_report(state)

    def worker(self, worker_id: str) -> WorkerReport:
        return _worker_report(self._worker(worker_id))

    def report(self) -> ReleaseReport:
        stage = None
        if self._rollout and self._rollout.stage_index >= 0:
            stage = self._rollout.stages[self._rollout.stage_index]
        return ReleaseReport(
            self._revision,
            self._active.release_id,
            self._candidate.release_id if self._candidate else None,
            self._previous.release_id if self._previous else None,
            self._rollout.state if self._rollout else self._terminal_state,
            stage.name if stage else None,
            stage.traffic_percent if stage else 0,
            self._ready_workers(self._active.release_id),
            self._ready_workers(self._candidate.release_id) if self._candidate else 0,
            sum(worker.status == "draining" for worker in self._workers.values()),
            sum(worker.active_work_count for worker in self._workers.values()),
        )

    def _validate_release_compatibility(
        self, active: ReleaseManifest, candidate: ReleaseManifest
    ) -> None:
        if active.release_id == candidate.release_id:
            raise ReleaseCompatibilityError("release id must change")
        if not _ranges_overlap(active.protocol, candidate.protocol):
            raise ReleaseCompatibilityError("protocol ranges do not overlap")
        if active.write_schema_version not in candidate.readable_schema_versions:
            raise ReleaseCompatibilityError("candidate cannot read active write schema")
        if candidate.write_schema_version not in active.readable_schema_versions:
            raise ReleaseCompatibilityError(
                "rollback release cannot read candidate write schema"
            )
        if active.policy_revision != candidate.policy_revision:
            raise ReleaseCompatibilityError(
                "binary canary cannot split policy revision"
            )

    def _manifest_for(self, release_id: str) -> ReleaseManifest:
        for manifest in (self._active, self._candidate, self._previous):
            if manifest and manifest.release_id == release_id:
                return manifest
        raise WorkerCompatibilityError(f"unknown release: {release_id}")

    def _ready_workers(self, release_id: str) -> int:
        return sum(
            worker.registration.release_id == release_id and worker.status == "ready"
            for worker in self._workers.values()
        )

    def _mark_release_draining(self, release_id: str) -> None:
        for worker in self._workers.values():
            if worker.registration.release_id == release_id and worker.status == "ready":
                worker.status = "drained" if worker.active_work_count == 0 else "draining"

    def _worker(self, worker_id: str) -> _WorkerState:
        try:
            return self._workers[worker_id]
        except KeyError as error:
            raise WorkerStateError(f"unknown worker: {worker_id}") from error

    def _expect_revision(self, expected: int) -> None:
        if expected != self._revision:
            raise StaleReleaseRevisionError(
                f"stale release revision {expected}; current={self._revision}"
            )


def _validate_manifest(manifest: ReleaseManifest) -> None:
    _require_text(manifest.release_id, "release_id")
    _require_text(manifest.binary_version, "binary_version")
    _positive_int(manifest.protocol.minimum, "protocol.minimum")
    _positive_int(manifest.protocol.maximum, "protocol.maximum")
    if manifest.protocol.minimum > manifest.protocol.maximum:
        raise ValueError("protocol range is inverted")
    if not manifest.readable_schema_versions:
        raise ValueError("at least one readable schema is required")
    for version in manifest.readable_schema_versions:
        _positive_int(version, "readable schema")
    _positive_int(manifest.write_schema_version, "write_schema_version")
    _positive_int(manifest.policy_revision, "policy_revision")
    _positive_int(manifest.feature_revision, "feature_revision")
    if manifest.write_schema_version not in manifest.readable_schema_versions:
        raise ValueError("release must read its own write schema")


def _validate_worker(worker: WorkerRegistration, manifest: ReleaseManifest) -> None:
    _require_text(worker.worker_id, "worker_id")
    if not manifest.protocol.minimum <= worker.protocol_version <= manifest.protocol.maximum:
        raise WorkerCompatibilityError("worker protocol is outside release range")
    if worker.write_schema_version != manifest.write_schema_version:
        raise WorkerCompatibilityError("worker write schema does not match manifest")
    if not all(
        version in worker.readable_schema_versions
        for version in manifest.readable_schema_versions
    ):
        raise WorkerCompatibilityError("worker cannot read every manifest schema")
    if worker.policy_revision != manifest.policy_revision:
        raise WorkerCompatibilityError("worker policy revision does not match manifest")
    if worker.feature_revision != manifest.feature_revision:
        raise WorkerCompatibilityError("worker feature revision does not match manifest")


def _validate_stages(stages: tuple[RolloutStage, ...]) -> None:
    if not stages:
        raise ValueError("rollout needs at least one stage")
    previous = 0
    for stage in stages:
        _require_text(stage.name, "stage.name")
        if not previous < stage.traffic_percent <= 100:
            raise ValueError("traffic stages must increase to at most 100")
        _positive_int(stage.min_samples, "stage.min_samples")
        previous = stage.traffic_percent
    if previous != 100:
        raise ValueError("last rollout stage must be 100 percent")


def _validate_slo(slo: SloPolicy) -> None:
    _ratio(slo.min_success_rate, "min_success_rate")
    _ratio(slo.max_observer_drop_rate, "max_observer_drop_rate")
    _ratio(slo.max_unknown_cost_rate, "max_unknown_cost_rate")
    if slo.max_p95_latency_ms <= 0:
        raise ValueError("max_p95_latency_ms must be positive")


def _assert_slo(window: SliWindow, stage: RolloutStage, slo: SloPolicy) -> None:
    if (
        window.sample_count < stage.min_samples
        or window.success_rate < slo.min_success_rate
        or window.p95_latency_ms > slo.max_p95_latency_ms
        or window.observer_drop_rate > slo.max_observer_drop_rate
        or window.unknown_cost_rate > slo.max_unknown_cost_rate
    ):
        raise SloViolationError(f"SLO rejected rollout stage: {stage.name}")


def _ranges_overlap(left: ProtocolRange, right: ProtocolRange) -> bool:
    return left.minimum <= right.maximum and right.minimum <= left.maximum


def _worker_report(state: _WorkerState) -> WorkerReport:
    return WorkerReport(
        state.registration.worker_id,
        state.registration.release_id,
        state.status,
        state.active_work_count,
        state.registration.protocol_version,
        state.registration.policy_revision,
        state.registration.feature_revision,
    )


def _ratio(value: float, name: str) -> None:
    if not 0 <= value <= 1:
        raise ValueError(f"{name} must be in [0, 1]")


def _positive_int(value: int, name: str) -> None:
    if not isinstance(value, int) or value < 1:
        raise ValueError(f"{name} must be a positive integer")


def _require_text(value: str, name: str) -> None:
    if not value.strip():
        raise ValueError(f"{name} must not be empty")
